src/OpenIddict.Core/OpenIddictController.cs

```diff
@@ -86,7 +88,37 @@ public class OpenIddictController<TUser, TApplication> : Controller where TUser
                 });
             }
 
-            return View("Authorize", Tuple.Create(request, await Manager.GetDisplayNameAsync(application)));
+            // Get all scopes associated with the current application and ensure all scopes specified in the authorization request 
+            // belongs to the current application or fail with an error, note that we automatically strip out standard **openid** scopes.
+            var requestedScopes = request.GetCustomScopes();
+
+            // Get all application scopes
+            var applicationScopes = await Manager.GetScopesByApplicationAsync(application);
+
+            // Get all request scopes
+            var requestScopes = await Manager.GetAuthorizationRequesteScopesAsync(requestedScopes);
+            
+            // TODO Add a better/smarter check
+            if (!requestScopes.Intersect(applicationScopes).SequenceEqual(requestScopes)){
+                return View("Error", new OpenIdConnectMessage{
+                    Error = OpenIdConnectConstants.Errors.InvalidRequest,
+                    ErrorDescription = "One or more of the request scopes does not belong to the calling client application"
+                });
+            }
+
+            var scopeModels = new List<ScopeModel>();
+            foreach (var scope in requestScopes)
+            {
+                // TODO this is higly inefficient in case of db lookups, so we should avoid await inside the foreach!
+                scopeModels.Add(new ScopeModel
+                {
+                    ScopeId = await Manager.GetScopeIdAsync(scope),
+                    DisplayName = await Manager.GetScopeDisplayNameAsync(scope),
+                    Description = await Manager.GetScopeDescriptionAsync(scope),
```

Comment:
Yes, just 3 state machines (only for the foreach) but in case the lookup happens from somewhere else (e.g. using a custom ```IOpenIddictStore```) this could have a much larger performance impact.